### ISSUE TITLE
Bug: layer initialization (in case shift buttons initially pressed) with Logitech X56

### ISSUE THREAD

**Issue by 3noix** [3noix/UJPS#21]
Issue seen by Zypp Torlen from [Star Citizen forum](https://robertsspaceindustries.com/spectrum/community/SC/forum/50174/thread/universal-joystick-programming-software-ujps-c/1301291):

"So I've been digging around the code to see if there's any way I can cheat the system on this one, but I haven't been able to find a good solution that doesn't involve moving private members of AbstractProfile into protected.

The Logitech X56 Throttle has a mode-switch dial on the bottom-left. Each mode corresponds to 1 button (namely, 34, 35, and 36). While the dial is on a mode, the corresponding button is simply held down perpetually.

The problem is, when starting up UJPS and loading a profile, it doesn't seem to read the initial state of the device, and it won't realize that button 34 is pressed until it is released and pressed again. So, because I use these buttons for Up, Middle, and Down layers, the program actually starts in the default Middle layer, even though my mode switch is physically in the Up position.

Is it possible to get UJPS to read the starting state of the controller on load?"

**Comment by 3noix** [comment]
?? No problem, it works perfectly for me with this code:

```
void Profile::runFirstStep()
{
	this->registerLayerDim1(Layers::In, tmwj, TMWJ::H2U);
	MapButton(tmwj, TMWJ::TG1, {"o"}, vj1, VJOY::DX1);
	MapButton(tmwj, TMWJ::TG1, {"i"}, vj1, VJOY::DX2);
}
```

**Comment by Eearslya** [comment]
Not sure if this might be controller-specific? I ran almost the exact same code as you, and I can reproduce the issue consistently.

```cpp
void Profile::runFirstStep() {
	this->registerLayerDim1(Layers::In, m_pThrottle, THROTTLE::E);
	MapButton(m_pJoystick, JOYSTICK::TG1, {"o"}, m_pVJoy1, VJOY::DX1);
	MapButton(m_pJoystick, JOYSTICK::TG1, {"i"}, m_pVJoy1, VJOY::DX2);
}
```

* If I have `E` held down when I click the Play button on UJPS, then pull the trigger, it will send `DX1`.

* If I let go of `E` and press the trigger, it will send `DX1` as well.

* If I then hold `E` again and press the trigger, it will send `DX2` as expected.

It's worth noting that any other button press or axis update from the throttle will cause UJPS to update the throttle's state, and then it will recognize that `E` is held down and apply the state properly. So if I hold down `E`, click Play, then push the throttle lever forward, pulling the trigger will trigger `DX2`.

**Comment by 3noix** [comment]
Ok, I understand. I did not have the problem because I did it with my Warthog stick and it is not really possible to press TG1 and H2U buttons without moving the axes. I will look more into it tomorrow with my throttle.

**Comment by Eearslya** [comment]
Ah, yeah. The Warthog is a bit more sensitive.. It might be hard for you to test, because I'm fairly sure this does require that 2 devices are involved. Otherwise, if the layer modifier and the modified key are on the same device, then pressing the keybind will cause the update I mentioned before to happen.

**Comment by 3noix** [comment]
I have both Warthog stick and throttle, so I will have no problem if the shift button is on the throttle and the other one on the stick.

But just to warn you: there is a chance it is due to a limitation of the DirectInput lib. In this case, I will not be able to do better. Maybe I will start to look at it this evening, otherwise this week-end.

**Comment by 3noix** [comment]
I made another test today with the shift buttons on the throttle (a 3 stable positions switch) and the other button on the joystick (the first stage of the trigger). I copied the code below. I made the test with and without the 3 last lines:

```
void Profile::runFirstStep()
{
	vj1->resetReport();
	
	this->registerLayerDim2(Layers::Down,   tmwt, TMWT::BSB);
	this->registerLayerDim2(Layers::Middle, tmwt, TMWT::BSM);
	this->registerLayerDim2(Layers::Up,     tmwt, TMWT::BSF);
	MapButton(tmwj, TMWJ::TG1, {"d"}, vj1, VJOY::DX1);
	MapButton(tmwj, TMWJ::TG1, {"m"}, vj1, VJOY::DX2);
	MapButton(tmwj, TMWJ::TG1, {"u"}, vj1, VJOY::DX3);
	
	vj1->setButton(VJOY::DX1, tmwj->buttonPressed(TMWJ::TG1) && tmwt->buttonPressed(TMWT::BSB));
	vj1->setButton(VJOY::DX2, tmwj->buttonPressed(TMWJ::TG1) && tmwt->buttonPressed(TMWT::BSM));
	vj1->setButton(VJOY::DX3, tmwj->buttonPressed(TMWJ::TG1) && tmwt->buttonPressed(TMWT::BSF));
}
```

I noticed a little issue (I will talk about it in the next post), but I was not able to reproduce your issue: whatever the position of the switch is when I press play (BSB, BSM or BSF), the selected layer is the good one... and I did not touch the throttle!

As the 3 positions of your dial are seen as 3 buttons by DirectInput, could you please confirm using the provided monitoring apps that the active button (34, 35 or 36) is not the good one when you start the monitoring app (always 35, the middle?)... to confirm it comes from the read of buttons positions, and not from the layer computation... even if the layer computation is so simple that I can't see where there could be an error.

If you confirm, it must mean it is only related to your X56. You can first check if there is a process related to your X56, and retry after killing it... otherwise we will have to check for info about such known issues with X56.

**Comment by 3noix** [comment]
So, a few words about the other issue I have seen when doing the test above: whatever the initial layer / switch position is, if I remove the 3 last line and if the trigger is pressed when I click "play", none of DX1, DX2 and DX3 buttons is pressed. It is in fact do to the fact that the actions does not manage the initialization.

The function VirtualJoystick::resetReport resets all buttons positions at false, all axes at 0 and all povs at center.

So the initialization must be done by calling vj1->setButton, vj1->setAxis or vj1->setPov... when calling MapButton, none of the DX1, DX2 and DX3 buttons is changed. And in fact, it is not an easy problem. Most of the time, it is easy:
- if you call MapButton, you certainly want to initialize the virtual button as the real button
- if you add a ActionButtonPulse, you will certainly want to init the virtual button at false

But UJPS allows you to use any Action (even your own), and with any Trigger (if you use a standard mapping with the AbstractProfile::Map), and in some cases, it is not possible to guess what the user wants, so the "manual initialization" will be still be useful. But it is indeed possible to provide a default initialization using a new AbstractAction::init function.

I open a new issue about that.

**Comment by 3noix** [comment]
There is one thing you can do to be sure... unfortunately I can't do the test for you as I don't have the X56: in StaticLibs/SOURCES/QtGameControllerModif/CODE/qgamecontroller_win.cpp:
- use breakpoints and a debugger, or simpler:
- outside of a function, add "#include <QtDebug>" or #include <iostream>
- and at the end, in the readGameController function add console outputs using "qDebug() << ...;" or std::cout to see why no signal is emitted... I think that GetDeviceState at line 257 must fail. And in this case, there must indeed be nothing to do.
And add "CONFIG += console" in "MonitoringX.pro"

**Comment by 3noix** [comment]
Hello. Did you try to look what happens into the readGameController function (where it returns from this function before any event)? For info, with the new versions you now have to look in the file "StaticLibs/SOURCES/QtGameControllerModif/CODE/GameControllerDirectInput.cpp"

**Comment by Eearslya** [comment]
Sorry, I got distracted from this project for a while. I made the modification you recommended; I added in a `qDebug()` to the button handler, like so:
```cpp
	//Buttons
	for (uint i=0; i<m_nbButtons; i++)
	{
		bool bPressed = (js.rgbButtons[i] & 0x80);
		if (bPressed != m_buttonsValues[i] || m_bFirstRead)
		{
			qDebug() << "New button value for " << i << ": " << bPressed;
			m_buttonsValues[i] = bPressed;
			emit gameControllerButtonEvent(new GameControllerButtonEvent{m_id,i,bPressed});
		}
	}
```

I also added 2 debug prints to check if either `Poll` or `GetDeviceState` fail.

```cpp
	hr = m_pJoystick->Poll();
	if (FAILED(hr))
	{
		qDebug() << "Poll failed";
		hr = m_pJoystick->Acquire();
		while (hr == DIERR_INPUTLOST) {hr = m_pJoystick->Acquire();}
		return;
	}
	if (FAILED(hr = m_pJoystick->GetDeviceState(sizeof(DIJOYSTATE2), &js))) {
		qDebug() << "GetDeviceState failed";
		return;
	}
```

I then launched `Monitoring0`. The console logs show that `Poll` fails 4 times initially (one for each device I have), and then all of them receive an update to all of their buttons at once. This one shows the Throttle receiving all of its updates.
![image](https://user-images.githubusercontent.com/2270586/43237467-c91dc9be-903e-11e8-922d-0958f5ef1f93.png)

In this case, Button 33 should have been `true`. Moving any axis or pushing any button on the throttle then shows `New button value for 33: true` as it's supposed to.

I never saw `GetDeviceState` fail anywhere in the log, just `Poll` failing once per device right at the start of the program.

I did some googling, and it appears I'm not alone, either. I've seen a few forum threads where people report that their device reports all buttons unpressed and all axes at 0 until they do something with DirectInput. Those threads are over 10 years old..so it might be safe to say that there is no resolution.
